src/Language/PureScript/Sugar/Names/Imports.hs

```diff
@@ -80,6 +87,22 @@ resolveModuleImport env ie (mn, imps) = foldM go ie imps
                    }
     resolveImport mn modExports ie'' impQual ss typ
 
+addPrimTypeErrorToEnvironment :: Environment -> Environment
+addPrimTypeErrorToEnvironment = augmentEnvironment primTypeErrorTypes primTypeErrorClasses
+
+addPrimRowToEnvironment :: Environment -> Environment
+addPrimRowToEnvironment = augmentEnvironment primRowTypes primRowClasses
+
+augmentEnvironment 
+  :: M.Map (Qualified (ProperName 'TypeName)) (Kind, TypeKind)
+  -> M.Map (Qualified (ProperName 'ClassName)) TypeClassData
+  -> Environment
+  -> Environment
+augmentEnvironment newTypes classes env = env
+  { typeClasses = typeClasses env `M.union` classes
+  , types = types env `M.union` newTypes
+  }
```

Comment:
This is where the magic happens. All of the `MonadState Environment` plumbing is necessary so that we can actually bring the `Prim.*` stuff into the `Environment` if we happen to import those modules.